File: lib/features/pokemon_list/data/repositories/pokemon_repository_impl.dart
import '../../../../core/domain/entities/pokemon.dart' as core;
import '../../../../core/domain/repositories/pokemon_repository.dart';
import '../../domain/entities/pokemon.dart' as feature;
import '../adapters/poke_api_adapter.dart';

class PokemonRepositoryImpl implements PokemonRepository {
  final PokeApiAdapter _adapter;

  PokemonRepositoryImpl(this._adapter);

  core.Pokemon _convertToCorePokemon(feature.Pokemon pokemon) {
    return core.Pokemon(
      id: int.tryParse(pokemon.id) ?? 0,
      name: pokemon.name,
      baseExperience: 0,
      height: pokemon.height,
      weight: pokemon.weight,
      isDefault: true,
      order: 0,
      sprites: core.Sprites(
        backDefault: '',
        backFemale: '',
        backShiny: '',
        backShinyFemale: '',
        frontDefault: '',
        frontFemale: '',
        frontShiny: '',
        frontShinyFemale: '',
        other: core.Other(
          dreamWorld: core.DreamWorld(
            frontDefault: '',
            frontFemale: '',
          ),
          home: core.Home(
            frontDefault: '',
            frontFemale: '',
            frontShiny: '',
            frontShinyFemale: '',
          ),
          officialArtwork: core.OfficialArtwork(
            frontDefault: pokemon.imageUrl,
            frontShiny: '',
          ),
        ),
      ),
      types: pokemon.types
          .map((type) => core.Type(
                slot: 1,
                type: core.Species(
                  name: type,
                  url: '',
                ),
              ))
          .toList(),
      abilities: pokemon.abilities
          .map((ability) => core.Ability(
                ability: core.Species(
                  name: ability,
                  url: '',
                ),
                isHidden: false,
                slot: 1,
              ))
          .toList(),
      stats: pokemon.stats.entries
          .map((e) => core.Stat(
                baseStat: e.value,
                effort: 0,
                stat: core.Species(
                  name: e.key,
                  url: '',
                ),
              ))
          .toList(),
      moves: pokemon.moves
          .map((move) => core.Move(
                move: core.Species(
                  name: move['name'] as String? ?? '',
                  url: '',
                ),
                versionGroupDetails: [],
              ))
          .toList(),
      forms: [],
      gameIndices: [],
      heldItems: [],
      locationAreaEncounters: '',
      cries: core.Cries(
        latest: '',
        legacy: '',
      ),
      pastAbilities: [],
      pastTypes: [],
      species: core.Species(
        name: pokemon.name,
        url: '',
      ),
    );
  }

  @override
  Future<List<core.Pokemon>> getPokemons(
      {int offset = 0, int limit = 20}) async {
    final pokemons = await _adapter.getPokemons(offset: offset, limit: limit);
    return pokemons.map(_convertToCorePokemon).toList();
  }

  @override
  Future<core.Pokemon> getPokemonById(int id) async {
    final pokemon = await _adapter.getPokemonById(id);
    return _convertToCorePokemon(pokemon);
  }

  @override
  Future<List<core.Pokemon>> searchPokemons(String query) async {
    final pokemons = await _adapter.searchPokemons(query);
    return pokemons.map(_convertToCorePokemon).toList();
  }

  @override
  Future<List<core.Pokemon>> getPokemonsByType(String type) async {
    // TODO: Implement getPokemonsByType
    return [];
  }

  @override
  Future<List<core.Pokemon>> getPokemonsByAbility(String ability) async {
    // TODO: Implement getPokemonsByAbility
    return [];
  }
}

File: bff/src/infrastructure/repositories/pokemon.repository.impl.ts
import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { firstValueFrom } from 'rxjs';
import { Pokemon } from '../../domain/entities/pokemon.entity';
import { PokemonRepository } from '../../domain/repositories/pokemon.repository';

@Injectable()
export class PokemonRepositoryImpl implements PokemonRepository {
  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {}

  private get baseUrl(): string {
    return this.configService.get<string>('POKEAPI_BASE_URL');
  }

  async getPokemonList(limit: number, offset: number): Promise<Pokemon[]> {
    const response = await firstValueFrom(
      this.httpService.get(`${this.baseUrl}/pokemon?limit=${limit}&offset=${offset}`),
    );

    const pokemonPromises = response.data.results.map(async (result: any) => {
      const pokemonResponse = await firstValueFrom(this.httpService.get(result.url));
      return this.mapToPokemon(pokemonResponse.data);
    });

    return Promise.all(pokemonPromises);
  }

  async getPokemonById(id: number): Promise<Pokemon> {
    const response = await firstValueFrom(this.httpService.get(`${this.baseUrl}/pokemon/${id}`));
    return this.mapToPokemon(response.data);
  }

  async searchPokemon(query: string): Promise<Pokemon[]> {
    const response = await firstValueFrom(
      this.httpService.get(`${this.baseUrl}/pokemon?limit=1000`),
    );

    const filteredResults = response.data.results.filter((result: any) =>
      result.name.toLowerCase().includes(query.toLowerCase()),
    );

    const pokemonPromises = filteredResults.map(async (result: any) => {
      const pokemonResponse = await firstValueFrom(this.httpService.get(result.url));
      return this.mapToPokemon(pokemonResponse.data);
    });

    return Promise.all(pokemonPromises);
  }

  async getPokemonsByType(type: string): Promise<Pokemon[]> {
    const response = await firstValueFrom(this.httpService.get(`${this.baseUrl}/type/${type}`));

    const pokemonPromises = response.data.pokemon.map(async (result: any) => {
      const pokemonResponse = await firstValueFrom(this.httpService.get(result.pokemon.url));
      return this.mapToPokemon(pokemonResponse.data);
    });

    return Promise.all(pokemonPromises);
  }

  async getPokemonsByAbility(ability: string): Promise<Pokemon[]> {
    const response = await firstValueFrom(
      this.httpService.get(`${this.baseUrl}/ability/${ability}`),
    );

    const pokemonPromises = response.data.pokemon.map(async (result: any) => {
      const pokemonResponse = await firstValueFrom(this.httpService.get(result.pokemon.url));
      return this.mapToPokemon(pokemonResponse.data);
    });

    return Promise.all(pokemonPromises);
  }

  private mapToPokemon(data: any): Pokemon {
    return {
      id: data.id,
      name: data.name,
      baseExperience: data.base_experience,
      height: data.height,
      weight: data.weight,
      sprites: {
        frontDefault: data.sprites.front_default,
        backDefault: data.sprites.back_default,
        frontShiny: data.sprites.front_shiny,
        backShiny: data.sprites.back_shiny,
      },
      types: data.types.map((type: any) => ({
        slot: type.slot,
        type: {
          name: type.type.name,
          url: type.type.url,
        },
      })),
      abilities: data.abilities.map((ability: any) => ({
        ability: {
          name: ability.ability.name,
          url: ability.ability.url,
        },
        isHidden: ability.is_hidden,
        slot: ability.slot,
      })),
      stats: data.stats.map((stat: any) => ({
        baseStat: stat.base_stat,
        effort: stat.effort,
        stat: {
          name: stat.stat.name,
          url: stat.stat.url,
        },
      })),
      moves: data.moves.map((move: any) => ({
        move: {
          name: move.move.name,
          url: move.move.url,
        },
        versionGroupDetails: move.version_group_details.map((detail: any) => ({
          levelLearnedAt: detail.level_learned_at,
          moveLearnMethod: {
            name: detail.move_learn_method.name,
            url: detail.move_learn_method.url,
          },
          versionGroup: {
            name: detail.version_group.name,
            url: detail.version_group.url,
          },
        })),
      })),
      species: {
        name: data.species.name,
        url: data.species.url,
      },
    };
  }
}
